### PR TITLE
Distributed-parallel-lazy

### PR DESCRIPTION
**Pull request by matthiasdiener**
Needs:
- [ ] https://github.com/inducer/pytato/pull/148
- [ ] https://github.com/inducer/grudge/pull/183
- [ ] https://github.com/illinois-ceesd/mirgecom/pull/562

---

To install/run via emirge:

```bash
$ ./install.sh --branch=production+distributed-parallel-lazy
$ source config/activate_env.sh
$ git clone --branch distributed-lazy https://github.com/illinois-ceesd/drivers_y2-isolator
$ cd drivers_y2-isolator/smoke_test
$ export PYOPENCL_CTX=0:1  # select GPU (will likely fail currently)
$ mpirun -n 2 python -m mpi4py isolator.py --lazy
```

### CODE REVIEW

**Comment by matthiasdiener** [comment]
@inducer: this works ok so far, however the health checks and time step calculations use global reductions, which distributed-lazy doesn't handle as far as I can see. I'm not sure what the best way to solve this is. 

I've written up a few comments here: https://notes.tiker.net/wkKBvXD_SD-meRPvmBUSaA

**Comment by MTCam** [comment]
> however the health checks and time step calculations use global reductions, which distributed-lazy doesn't handle as far as I can see. 

Note that these are not the only mpi reductions in the driver.  There are more in this routine:

https://github.com/illinois-ceesd/drivers_y2-isolator/blob/d443de1bf2848acdf4b96db57a3c82b2852db5af/isolator.py#L1239

And also these that look even more like the ones in Mike's health check:
https://github.com/illinois-ceesd/drivers_y2-isolator/blob/d443de1bf2848acdf4b96db57a3c82b2852db5af/isolator.py#L1107

Both of these routines are called every step.  It is confusing to me how the ones in health check would fail when these seem to pass without issue.

**Comment by matthiasdiener** [comment]
It turns out eager with multiple ranks also fails with the same error (see e.g. https://github.com/illinois-ceesd/drivers_y2-isolator/runs/4809937376)

**Comment by MTCam** [comment]
> It turns out eager with multiple ranks also fails with the same error (see e.g. [illinois-ceesd/drivers_y2-isolator/runs/4809937376](https://github.com/illinois-ceesd/drivers_y2-isolator/runs/4809937376)), so reductions might not be an issue after all.

The issue with eager is fixed by hard-coding the throat location and height in the isolator initialization routines, instead of "discovering" it by examining the mesh coordinates.  I have fixed this in my local copy and it now passes its own internal fidelity test on 4 MPI ranks.  CC: @anderson2981

**Comment by matthiasdiener** [comment]
This diff by MikeC also fixes distributed-lazy

```diff
diff --git a/isolator.py b/isolator.py
index 7e55353..ae4a5cd 100644
--- a/isolator.py
+++ b/isolator.py
@@ -394,6 +394,9 @@ class InitACTII:
                 throat_height = ytop_flat[inode] - ybottom_flat[inode]
                 throat_loc = xpos_flat[inode]

+        throat_height = self._throat_height
+        throat_loc = self._x_throat
+
         #print(f"throat height {throat_height}")
         for inode in range(xpos_flat.size):
             area_ratio = (ytop_flat[inode] - ybottom_flat[inode])/throat_height
```

File: isolator.py [hunk 1]
```diff
@@ -102,15 +102,15 @@ def filter(self, record):
...
-# root_logger.addHandler(h2)
+h1 = logging.StreamHandler(sys.stdout)
+f1 = SingleLevelFilter(logging.INFO, False)
+# h1.addFilter(f1)
+root_logger = logging.getLogger()
+root_logger.addHandler(h1)
+h2 = logging.StreamHandler(sys.stderr)
+f2 = SingleLevelFilter(logging.INFO, True)
+# h2.addFilter(f2)
+root_logger.addHandler(h2)
```
Comment (MTCam) [inline isolator.py:113]: This undoes the disabling of this by Matthias in ([71f2813](https://github.com/illinois-ceesd/drivers_y2-isolator/pull/6/commits/71f2813097b059c8858a5f3311699a0ca98d786a).  I suspect he needs this block disabled to effectively debug.

**Comment by anderson2981** [comment]
I'm merging this to bring all our changes into main for the driver. this became my development head and has lots of un-related additions.

**Pull request merged**